API: libpacemaker: Move crm_rule functionality to libpacemaker

nrwahl2 opened the pull request:
Closes T58

nrwahl2 commented on lib/pacemaker/pcmk_rule.c:
```diff
@@ -0,0 +1,290 @@
+ */
+int
+pcmk__check_rules(pcmk__output_t *out, xmlNodePtr input, crm_time_t *date,
+                  const char **rule_ids)
```
I think it makes more sense for the API functions should handle just one rule at a time, instead of processing the whole list at once as `crm_rule` does. That way, we get to keep the return code for every rule we check, instead of just "success or the last non-success return code."

For this draft, I wrote it to accept a list of rule IDs like the existing logic in `crm_rule`. That way the output structure doesn't change. This can easily be updated to take just one rule ID though.

nrwahl2 replied:
Updated to add static wrappers for calling with a single rule ID.

nrwahl2 commented on cts/cli/regression.rules.exp:
```diff
@@ -11,23 +11,17 @@ crm_rule: --check requires use of --rule=
 =#=#=#= End test: crm_rule given no rule to check - Incorrect usage (64) =#=#=#=
 * Passed: crm_rule       - crm_rule given no rule to check
 =#=#=#= Begin test: crm_rule given invalid input XML =#=#=#=
-(log_xmllib_err@xml.c:891) 	error: XML Error: Entity: line 1: parser error : Start tag expected, '<' not found
```
Do we need this? As mentioned in the commit message, we can get this (minus the `"@xml.c:891"`) if add verbosity, but we no longer get it by default.

nrwahl2 replied:
This is actually what caused the CI build to fail.

`build-all-voting` succeeded, because `cts-cli` doesn't produce these `(log_xmllib_err@xml.c:891)` messages, which I removed from the test output.

However, `build-rpms` failed. The `(log_xmllib_err@xml.c:891)` messages DO appear in the output when running `make -C rpm rpm`.

I don't know what could be causing this discrepancy.

nrwahl2 replied:
Okay, here's the deal... Initially, I ran `cts-cli -t rules` after writing the code in this PR, and it didn't include the `log_xmllib_err` lines. So I removed them from `regression.tools.exp`. All CTS tests ran fine on my machine after that, and I pushed.

CI passed `build-all-voting` and failed `build-rpms`. I reproduced in my machine that `make -C rpm rpm` fails due to a diff in the `cts-cli` tests.

It's fixed now after I put those lines back in. It turns out there's a really strange issue with `cts-cli`, where it prints the `log_xmllib_err` lines with certain combinations of tests but not others. Likewise, another test group (`tools` IIRC) `unpack_resources` errors with some test group combinations and not with others. I filed https://projects.clusterlabs.org/T548 for this issue. The best bet is to run the full `cts-cli` suite, to get the same output as the regression.*.exp files and to update them when needed.

nrwahl2 commented on tools/crm_rule.c:
```diff
@@ -87,175 +82,6 @@ mode_cb(const gchar *option_name, const gchar *optarg, gpointer data, GError **e
-     * double checking can't hurt.
-     */
-    CRM_ASSERT(match != NULL);
-    CRM_ASSERT(find_expression_type(match) == time_expr);
```
`time_expr` was not defined anywhere so I dropped it in when moving to libpacemaker

kgaillot replied:
? It's in include/crm/pengine/rules.h

nrwahl2 replied:
Ouch, yeah thank you. I saw this below, with `time_expr` in the doxygen block of the helper function when the argument was really called `expr`, and blanked out. It wouldn't have compiled if `time_expr` weren't defined.
```
 * \param[in]  time_expr    date_expression XML
...
 */
static int
eval_date_expression(xmlNode *expr, crm_time_t *now, crm_time_t *next_change)
```

Will fix.

nrwahl2 commented:
@clumens Can you review when you get a chance?

nrwahl2 commented on lib/pacemaker/pcmk_rule.c:
```diff
@@ -0,0 +1,290 @@
+ * \brief Initialize the cluster working set for checking rules
+ *
+ * Make our own copies of the CIB XML and date/time object, if they're not
+ * \c NULL. This way we don't have to take ownership of the objects passed via
```
seems simpler for callers to deal with this way

nrwahl2 commented on include/pacemaker.h:
```diff
@@ -186,6 +186,20 @@ int pcmk_list_nodes(xmlNodePtr *xml, char *node_types);
+ *
+ * \return Standard Pacemaker return code
+ */
+int pcmk_check_rules(xmlNodePtr *xml, xmlNodePtr input, crm_time_t *date,
```
Not sure why these are getting documented in the header file when we've typically put doxygen comments in the code, but I went ahead and followed the pattern.

kgaillot replied:
Either is acceptable. Documenting in the header is easier for external developers who might think to look to the headers for documentation rather than the online API docs; documenting in the source is easier for us when reading code.

nrwahl2 commented:
Feature set bump?

nrwahl2 commented on lib/pacemaker/pcmk_output.c:
```diff
@@ -1864,6 +1864,50 @@ attribute_xml(pcmk__output_t *out, va_list args)
+    pcmk__output_create_xml_node(out, "rule-check",
+                                 "rule-id", rule_id,
+                                 "rc", rc_str,
+                                 NULL);
```
The messages from rule_check_default could be added to the XML for additional clarity. Not sure if that's desirable and whether that would be allowable in the context of schema compatibility.

kgaillot replied:
Not worth the effort at this point

nrwahl2 replied:
With the latest update, this seems trivial to implement if we wanted to do so. We could add an optional `message` attribute and set it to a known message or something like `"Error occurred checking rule"`. It could go in a separate PR if we wanted it, since this one is already unwieldy.

nrwahl2 commented on lib/pacemaker/pcmk_rule.c:
```diff
@@ -0,0 +1,290 @@
+            rc = ENXIO;
+            out->err(out, "Rule either must not use date_spec, or use "
+                     "date_spec with years= but not moon=");
+            goto done;
```
We could potentially still call `out->message()` in this "unsupported" case, and add a handler to the format functions for `ENXIO`.

Maybe even do the same for the other cases where we bail out early (no rule found, more than one rule found, not exactly one date expression).

clumens replied:
I don't really have any strong feelings on that one way or the other.  Maybe try converting one piece and see what it looks like.

nrwahl2 replied:
Looks fine to me. I'll point it out in the new regression test output in a sec. I don't have strong feelings here either.

kgaillot commented on lib/pacemaker/pcmk_rule.c:
```diff
@@ -0,0 +1,291 @@
+// Documented in pacemaker.h
+int
+pcmk_check_rules(xmlNodePtr *xml, xmlNodePtr input, crm_time_t *date,
+                 const char **rule_ids)
```
The public API should be added in its own commit for the message

nrwahl2 replied:
Done

kgaillot commented on tools/Makefile.am:
```diff
@@ -139,7 +139,8 @@ crm_rule_SOURCES 	= crm_rule.c
 			  $(top_builddir)/lib/pengine/libpe_status.la   \
-			  $(top_builddir)/lib/common/libcrmcommon.la
+			  $(top_builddir)/lib/common/libcrmcommon.la	\
+			  $(top_builddir)/lib/pacemaker/libpacemaker.la
```
libraries are listed highest-to-lowest level left-to-right (really only has an impact when linking statically, but it's good habit)

nrwahl2 replied:
So libpacemaker first as the highest-level library? Not a pattern that's consistently followed so I wasn't aware of it. Easy switch.

kgaillot reviewed (commented):
Not a comprehensive review, just a first look

kgaillot commented:
> Feature set bump?

No, since the tool interface isn't changing

clumens commented on include/pacemaker.h:
```diff
@@ -186,6 +186,40 @@ int pcmk_list_nodes(xmlNodePtr *xml, char *node_types);
+ * \return Standard Pacemaker return code
+ */
+int pcmk_check_rules(xmlNodePtr *xml, xmlNodePtr input, crm_time_t *date,
+                     const char **rule_ids);
```
This function could also be written to take `...` for the list of rules.  I'd have to look at how it's used elsewhere, but it could potentially save some work at the callers since you wouldn't have to construct the data type to pass.

nrwahl2 replied:
It would be easier for some callers so I like it in theory. However, I'm not aware of a way to convert an existing array of strings to a list of separate arguments (or even a `va_list`), and searching isn't turning anything useful up. Closest search result I found said you can't: https://stackoverflow.com/a/14705987/7660197. Let me know if you have ideas that don't require having a variadic version and a non-variadic version and we can give it a shot

I figure at least accepting a char* array is clear and straightforward, even if it requires a little bit of extra typing sometimes.

clumens replied:
There's no way that I know of to do that.

nrwahl2 replied:
Let me know; we can do it either way. I think there's a reasonable possibility that a caller will want to check multiple rules at once and parse the results together. Maybe fairly uncommon though...

I'd be less concerned about efficiency since it's probably an uncommon need and maybe not super time-sensitive; and more concerned about clunkiness for a caller. I've been annoyed at REST APIs that only let me get a single item at a time before.

kgaillot replied:
Might as well leave it as it is

clumens commented on include/pcmki/pcmki_rule.h:
```diff
@@ -0,0 +1,31 @@
+#ifndef PCMK__PCMKI_PCMKI_RULE__H
```
Remember the copyright boilerplate up here, too.

nrwahl2 replied:
Thanks! I copied `include/pcmki/pcmki_cluster_queries.h` to use as my template, and it doesn't have copyright boilerplate. I'll add that before the next push.

clumens commented on lib/pacemaker/pcmk_rule.c:
```diff
@@ -0,0 +1,291 @@
+     * checking can't hurt.
+     */
+    CRM_ASSERT(match != NULL);
+
```
You dropped the check of `find_expression_type` here.

nrwahl2 replied:
Thank you, yeah, that was a brain fart. Intentional at the time but definitely wrong: https://github.com/ClusterLabs/pacemaker/pull/2836#discussion_r957663996

clumens commented on lib/pacemaker/pcmk_rule.c:
```diff
@@ -126,15 +126,15 @@ rule_init(pcmk__output_t *out, xmlNodePtr input, crm_time_t *date,
 static int
-crm_rule_check(pcmk__output_t *out, pe_working_set_t *data_set,
-               const char *rule_id, crm_time_t *date)
+do_check_rule(pcmk__output_t *out, pe_working_set_t *data_set,
```
I know @kgaillot isn't really a fan of `do_` functions.

nrwahl2 replied:
Fair enough I thought `crm_rule_check` looked too much like a public name, and a bare `rule_check` or `check_rule` could be confusing when we already have `pcmk__rule_check` and `pcmk_rule_check` (the inline wrappers).

kgaillot replied:
Maybe `check_rule_effectiveness` or `eval_rule`

nrwahl2 replied:
`eval_rule` seems like a good fit since `eval` is the pattern for this kind of thing in `rules.c` (e.g., `pe_eval_rules`, `pe_eval_expr`). It's almost too generic since we might want to use `eval_rule` in something like `rules.c` later -- but it's static so we can change it if we ever need to. (Static name conflicts won't break compilation, just good to avoid for logging purposes).

nrwahl2 commented on lib/pacemaker/pcmk_rule.c:
```diff
@@ -155,58 +155,58 @@ crm_rule_check(pcmk__output_t *out, pe_working_set_t *data_set,
                      "date_spec with years= but not moon=");
-            goto done;
+            freeXpathObject(xpath_obj);
+            return ENXIO;
```
Seems like `EOPNOTSUPP` might be better here too

nrwahl2 replied:
Done; will revert if we're concerned about anyone depending on `ENXIO` specifically in this situation

nrwahl2 commented:
Updated:
* Return `EOPNOTSUPP` instead of `ENXIO` when a rule contains a date_spec without `years` or with `years=` and `moon=`. This is something we don't support checking, not a rule not found situation.
* Added copyright to `pcmki_rule.h` and `pcmki_cluster_queries.h`.
* Separated public API additions into their own commit
* Moved libpacemaker to the beginning of `LDADD` for `crm_rule`
* Removed some redundant arguments in `eval_rule_expression()` and `do_check_rule()`. Use `data_set->now` after initialization so that we can drop an argument to those helpers and also make the `date` arguments const for the API functions.
* Re-added the `time_expr` line that I mistakenly dropped.
* Write a `"rule-check"` message when we fail to check a rule, so that we have a check result item for every rule that we request to check, even if we can't find it or don't support checking it.
* Added XML versions with validations to all `crm_rule` regression tests that were missing them.

I've got a small mistake in regression test output on this update, so I'm about to push again to fix that after `make -C rpm rpm` finishes on my machine to validate.

nrwahl2 commented on cts/cli/regression.rules.exp:
```diff
@@ -28,6 +28,7 @@ crm_rule: Couldn't parse input from STDIN
 * Passed: echo           - crm_rule given invalid input XML on stdin
 =#=#=#= Begin test: Try to check a rule that doesn't exist =#=#=#=
 No rule found with ID=blahblah
+Could not determine whether rule blahblah is expired
```
@clumens Example of new text output when we call `out->message` in the event of rule-check errors too.

kgaillot replied:
The text seems redundant, not sure what we should do here. The ideal would be something like

    Could not determine whether rule is expired: Rule blahblah not found

but that's probably not worth the hassle of figuring out

nrwahl2 replied:
I think it can be done without too much trouble, if we want to send the "Could not determine" message to `out->err` instead of `out->info` like we do with the other formatter messages.

kgaillot replied:
Seems like it should be an error anyway

nrwahl2 replied:
@kgaillot I've got it working so that it all prints on one line for "rule not found" and "more than one rule" cases for text output. The unsupported date format case would still have two lines, because we'd be returning `EOPNOTSUPP` for both subcases and the formatter has no way to differentiate.

The only caveat is that the error **messages** for "rule not found" and "more than one rule" won't show up in XML output. XML output would only get the return code. (XML output would still get the error messages for the unsupported date formats.)

That might be okay, as long as we use each return code for a unique situation: `ENXIO` for "rule not found" and `pcmk_rc_duplicate_id` for "more than one rule." I'm doing that in the next update regardless.

Thoughts?

```
Could not determine whether rule blahblah is expired: Rule not found
...
Could not determine whether rule duplicate-rule-id is expired: Found more than one rule with matching ID
...
Rule cli-prefer-rule-dummy-date_spec-years-moon must either not use date_spec, or use date_spec with years= but not moon=
Could not determine whether rule cli-prefer-rule-dummy-date_spec-years-moon is expired: Unsupported date format
...
Can't check rule cli-no-date_expression-rule because it does not have exactly one date_expression
Could not determine whether rule cli-no-date_expression-rule is expired: Unsupported date format
```

```
<pacemaker-result api-version="X" request="crm_rule -c -r blahblah --output-as=xml">
  <rule-check rule-id="blahblah" rc="105"/>
  <status code="105" message="No such object">
  <!-- this would go away
    <errors>
      <error>No rule found with ID=blahblah</error>
    </errors>
  -->
  </status>
</pacemaker-result>
```

nrwahl2 replied:
I'll go ahead and push it with the other changes. It's easy to modify if we decide we don't want this.

nrwahl2 commented on cts/cli/regression.rules.exp:
```diff
@@ -91,6 +92,7 @@ No rule found with ID=blahblah
 * Passed: crm_rule       - Try to check a rule that doesn't exist
 =#=#=#= Begin test: Try to check a rule that doesn't exist, with XML output =#=#=#=
 <pacemaker-result api-version="X" request="crm_rule -c -r blahblah --output-as=xml">
+  <rule-check rule-id="blahblah" rc="105"/>
```
@clumens Example of new XML output when we call out->message in the event of rule-check errors too.

One advantage is that I think previously, the status code would be the last error we found. Now we get a discrete rule-check element and return code for every rule we checked.

kgaillot replied:
That does seem helpful. Need to update the xml/api/crm_rule schema and bump PCMK__API_VERSION

nrwahl2 replied:
zeroOrMore to oneOrMore rule-checks? Can't think of what else might need to change

kgaillot replied:
New schemas should remain compatible with output produced by earlier versions, so zeroOrMore

nrwahl2 replied:
I'm saying I don't see what in the schema needs to be updated for this PR. (There is every chance in the world I've got a glaring blind spot right now.)

kgaillot replied:
My bad, I thought this was completely new but it's just a new occurrence

nrwahl2 replied:
Ah gotcha, no worries

clumens commented on cts/cli/regression.rules.exp:
```diff
@@ -145,13 +148,16 @@ Rule cli-prefer-rule-dummy-date_spec-only-years satisfies conditions
 * Passed: crm_rule       - Verify date_spec rule with years is in effect
 =#=#=#= Begin test: Try to check a rule whose date_spec does not contain years= =#=#=#=
 Rule must either not use date_spec, or use date_spec with years= but not moon=
+Could not determine whether rule cli-prefer-rule-dummy-date_spec-without-years is expired
```
Looking at these two next to each other, it'd be nice if the top message also said "Rule cli-prefer-rule-dummy-date_spec-without-years must either...".

nrwahl2 replied:
Good point. The other three "failed to check rule" messages all include the rule ID.

kgaillot commented on lib/pacemaker/pcmk_rule.c:
```diff
@@ -152,11 +152,13 @@ do_check_rule(pcmk__output_t *out, pe_working_set_t *data_set,
+            out->err(out, "More than one rule with ID=%s found", rule_id);
+        }
+        out->message(out, "rule-check", rule_id, ENXIO);
         return ENXIO;
```
We have pcmk_rc_multiple but unfortunately that's specifically for "Resource active on multiple nodes". I think this sort of thing comes up enough to be worth a more generic code like pcmk_rc_duplicate_id.

nrwahl2 replied:
agreed, done

kgaillot reviewed (commented):
nice work

nrwahl2 commented on lib/pacemaker/pcmk_output.c:
```diff
@@ -1864,6 +1864,50 @@ attribute_xml(pcmk__output_t *out, va_list args)
+    } else {
+        out->info(out, "Could not determine whether rule %s is expired",
+                  rule_id);
+    }
```
Yep, will do

nrwahl2 commented on lib/pacemaker/pcmk_rule.c:
```diff
@@ -0,0 +1,272 @@
+
+    // Make our own copy of the given input or fetch the CIB and use that
+    if (input != NULL) {
+        new_data_set->input = copy_xml(input);
```
I noticed that, starting to look through the libxml2 docs recently... kind of frustrating for functions that clearly don't modify anything

nrwahl2 commented on lib/pacemaker/pcmk_rule.c:
```diff
@@ -0,0 +1,272 @@
+    int max = 0;
+
+    CRM_ASSERT((out != NULL) && (data_set != NULL) && (rule_id != NULL)
+               && (date != NULL));
```
Fair enough, I'm just paranoid. Will remove it.

nrwahl2 commented on lib/pacemaker/pcmk_rule.c:
```diff
@@ -0,0 +1,272 @@
+     * We do this in steps to provide better error messages. First, check that
+     * there's any rule with the given ID.
+     */
+    xpath = crm_strdup_printf("//rule[@id='%s']", rule_id);
```
Oh yeah I'd prefer that. I didn't think to check for it.

nrwahl2 commented:
* Rename do_check_rule() to eval_rule(), rule_init to init_rule_check()
* Use `XML_TAG_RULE`
* Add rule ID to `"moon="` error message, like the others
* Remove asserts from static functions
* Use switch in `rule_check_default()` and return the value of `out->info()` to make it compact
* Add `pcmk_rc_duplicate_id` and use it when more than one matching rule is found
* Put "rule not found" and "more than one matching rule found" messages on same line as "Could not determine..." (but drop them from XML output) (up for discussion of whether we want this since it drops "no rule" and "multiple rules" from XML error output -- https://github.com/ClusterLabs/pacemaker/pull/2836#discussion_r958952687)

kgaillot commented on lib/pacemaker/pcmk_rule.c:
```diff
@@ -111,6 +111,8 @@ init_rule_check(pcmk__output_t *out, xmlNodePtr input, crm_time_t *date,
     return pcmk_rc_ok;
 }
 
+#define XPATH_NODE_RULE "//" XML_TAG_RULE "[@id='%s']"
```
Some of the other xpath components have constants e.g. XML_ATTR_ID, XML_EXPR_ATTR_OPERATION

nrwahl2 replied:
I must have stopped checking after I saw that `date_expression` didn't have a constant

kgaillot commented on include/crm/common/results.h:
```diff
@@ -108,6 +108,7 @@ enum pcmk_rc_e {
     /* When adding new values, use consecutively lower numbers, update the array
      * in lib/common/results.c, and test with crm_error.
      */
+    pcmk_rc_duplicate_id        = -1033,
```
This commit message would be API:

nrwahl2 replied:
Makes sense, wish I'd caught that in 73a46d1a :)
Will fix

kgaillot commented on lib/pacemaker/pcmk_output.c:
```diff
@@ -1884,10 +1884,30 @@ rule_check_default(pcmk__output_t *out, va_list args)
+     * because we don't want a newline.
+     */
+    fprintf(stderr, "Could not determine whether rule %s is expired: ",
+            rule_id);
```
For future-proofing (vs new or reimplemented formats) I'd rather just duplicate this in the out->err() calls. That also lets us merge the switches and simplify the default

nrwahl2 replied:
Good point

kgaillot commented on cts/cli/regression.rules.exp:
```diff
@@ -91,16 +91,14 @@ No rule found with ID=blahblah
-    <errors>
-      <error>No rule found with ID=blahblah</error>
-    </errors>
-  </status>
```
I hate losing detail in the XML format because we're pushing everyone to use it, and I don't want to give them a reason not to. We can call the same out->err()'s in the XML implementation

nrwahl2 replied:
Good point. At that point I might as well move the `out->err()` calls back to `eval_rule()` in `pcmk_rule.c` so that they printed for all formats. Then remove the default error from the `rule_check_default` formatter.

kgaillot reviewed (commented):
These are nice improvements

nrwahl2 commented:
Updated to address review.
* Use `XML_ATTR_ID` and `XML_EXPR_ATTR_OPERATION`. I don't see any more that have defined constants.
* Update `pcmk_rc_duplicate_id` commit message to "API:"
* Re-did the error handling again. I appreciate the push to reorganize it. Whether we stick with this approach or not, I like it a lot better.
  * Now `eval_rule()` doesn't have any side effects. We just pass the rc and error message (via pointer) back to the caller let the caller call `out->message()`.
  * `out->message()` now accepts an error argument. The default case uses the error if it's not `NULL`, or "unexpected error" if it is `NULL`.
  * The XML formatter also calls `out->err()` in the event of an error.

nrwahl2 commented on lib/pacemaker/pcmk_output.c:
```diff
@@ -1864,6 +1864,65 @@ attribute_xml(pcmk__output_t *out, va_list args)
+        case pcmk_rc_within_range:
+            return out->info(out, "Rule %s is still in effect", rule_id);
+        case pcmk_rc_ok:
+            return out->info(out, "Rule %s satisfies conditions", rule_id);
```
I don't think this is very intuitive... for text output, I wonder if should we combine both `pcmk_rc_within_range` and `pcmk_rc_ok` into one case for "Rule %s is in effect". Or something like `"Rule %s is in effect: within date range"` and `"Rule %s is in effect: satisfies cron conditions"`

kgaillot replied:
I wouldn't bother at this point

nrwahl2 commented on lib/pacemaker/pcmk_output.c:
```diff
@@ -1864,6 +1864,65 @@ attribute_xml(pcmk__output_t *out, va_list args)
+    int result = va_arg(args, int);
+    const char *error = va_arg(args, const char *);
+
+    char *rc_str = pcmk__itoa(pcmk_rc2exitc(result));
```
I wonder if we should map both success codes (`pcmk_rc_ok` and `pcmk_rc_within_range`) to `pcmk_rc_ok`, to make things simpler for the caller. An rc of 0 would mean "the rule is in effect" -- instead of checking for two different codes. I don't think it would deprive anyone of useful information.

Not sure if we really want `pcmk_rc2exitc` here but I'm leaving it alone unless requested.

kgaillot commented on lib/pacemaker/pcmk_rule.c:
```diff
@@ -0,0 +1,296 @@
+    }
+
+    if (num_results > 1) {
+        // Should not be possible; schema prevents this
```
One gotcha that's easy to forget is that schema validation can be disabled, and then the input can be anything as long as it's legal XML. So while the schema validation is important for the user experience, it can't be relied on by the code.

kgaillot reviewed (approved):
Stick a fork in it!
